content/4_Gold/Digit_DP.mdx

```diff
@@ -0,0 +1,146 @@
+---
+id: digit-dp
+title: 'Digit DP'
+author: Jesse Choe
+prerequisites:
+  - intro-dp
+description: 'Finding the number of integers in a range that have a property.'
+frequency: 1
+---
+
+<FocusProblem problem='focus' />
+
+## General Resources
+
+<Resources>
+	<Resource
+		source='AR'
+		title='Dynamic Programming for Computing Contests'
+		url='https://dp-book.com/Dynamic_Programming.pdf'
+	>
+
+	</Resource>
+	<Resource
+		source="GFG"
+		title="Digit DP | Introduction"
+		url="https://www.geeksforgeeks.org/digit-dp-introduction/"
+	>
+
+	</Resource>
+	<Resource
+		source="Youtube"
+		title="Introduction to Digit Dynamic Programming"
+		url="https://www.youtube.com/watch?v=heUFId6Qd1A"
+	>
+		Great introduction video
+	</Resource>
+	<Resource
+		source="CF"
+		title="Digit DP"
+		url="https://codeforces.com/blog/entry/53960"
+	>
+
+	</Resource>
+</Resources>
+
+## Solution - Odometer
+
+### Without Dynamic Programming
+
+We can solve this problem in $\mathcal{O}(Y - X)$ time by looping through $X \ldots Y$ and checking whether a given number is an interesting number. However, since $Y - X$ can be large, the naive solution is too slow, requiring us to optimize it.
+
+### With Dynamic Programming
+
+One way to optimize the brute force approach of looping from $X$ to $Y$ and counting interesting numbers is by using dynamic programming. The DP approach involves considering all 9 digits one at a time to occupy at least half of the number.
+
+Let $\texttt{dp[i][und][k][is0]}$ where $i$ is the current position, $und$ is whether you have gone below the actual number, $k$ is a counter to see if you have at least half of the same digit, and $is0$ is a boolean if you still have leading zeros. The transition involves looping through 9 digits to add to the current location and comparing with the digit we want to occupy at least half of the number.
+
+This algorithm is fast enough as the number of digits is small and there are only nine digits.
+
+<LanguageSection>
+<CPPSection>
+
+Code from
+[here](https://raw.githubusercontent.com/thecodingwizard/competitive-programming/master/USACO/2014open/silver/odometer.cpp).
```

Comment:
better as "Code by Nathan Wang", and add link to github instead of just the raw file